### ISSUE TITLE
Binning Incomes

### ISSUE THREAD

**Issue by hayleefay** [hayleefay/Tax-Calculator#1]
@rickecon, as of now the binning for incomes is split as follows:
![image](https://user-images.githubusercontent.com/12599305/32564759-d4c9a304-c47a-11e7-9d81-a4a80fde541c.png)

Would you like me to use the first bin (showing incomes below 0) or omit it?

**Comment by rickecon** [comment]
Let's combine it with the 2nd bin so we have `[-9e+99, 9999.0]`.

**Comment by hayleefay** [comment]
@rickecon, I will do that. Here is just a quick mock-up to give an idea. I still need to clean it up.
![image](https://user-images.githubusercontent.com/12599305/32566630-25c89ca8-c47e-11e7-9efe-a162d16dc222.png)

**Comment by rickecon** [comment]
Yes. That looks really good. I can see some things we'll need to change.
1. Label the `x`-axis as "Unadjusted gross income". I think we can get rid of the "($thousands)" because we have the "K" in the category labels.
2. Change the `y`-axis labels of the bottom plot to be in thousands. Let's try the "K notation there also and see how it looks. Otherwise, we could put a little parenthetical "(000s)" in the axis label.
3. In the bottom plot, instead of the `y`-axis labels of "Tax units with tax increase" and "Tax units with tax cut", it might make more sense to put those two descriptions as a legend with "orange = Tax units with tax increase" and "gray = Tax units with tax cut".
  a. If we do this, we could leave the `y`-axis label blank if we use the "K" notation in the axis tick lables.
  b. Or we could put one `y`-axis label of (thousands of tax units).
4. The negative values in the `y`-axis of the bottom figure are confusing. Let's include a heavier black line at zero then label the negative values as positives, even though they are really negative in the data (I think).
5. The top plot is misleading. It makes the TJCA look like a huge cut for the wealthy. It may still be, but a more informative plot would be "Difference in tax liability as a percent of total income (effective tax rate)".
  a. This change would require the `y`-axis label of the top plot to be "Diff. in percentage points".

**Comment by hayleefay** [comment]
@rickecon For the quantitative note, do you want to big blue year on the chart still? Do you want anything else altered? Also, should I add a unit on the Average Tax Chart?

**Comment by rickecon** [comment]
Yes. That looks really good. I can see some things we'll need to change.
1. Label the `x`-axis as "Unadjusted gross income". I think we can get rid of the "($thousands)" because we have the "K" in the category labels.
2. Change the `y`-axis labels of the bottom plot to be in thousands. Let's try the "K notation there also and see how it looks. Otherwise, we could put a little parenthetical "(000s)" in the axis label.
3. In the bottom plot, instead of the `y`-axis labels of "Tax units with tax increase" and "Tax units with tax cut", it might make more sense to put those two descriptions as a legend with "orange = Tax units with tax increase" and "gray = Tax units with tax cut".
  a. If we do this, we could leave the `y`-axis label blank if we use the "K" notation in the axis tick lables.
  b. Or we could put one `y`-axis label of (thousands of tax units).
4. The negative values in the `y`-axis of the bottom figure are confusing. Let's include a heavier black line at zero then label the negative values as positives, even though they are really negative in the data (I think).
5. The top plot is misleading. It makes the TJCA look like a huge cut for the wealthy. It may still be, but a more informative plot would be "Difference in tax liability as a percent of total income (effective tax rate)".
  a. This change would require the `y`-axis label of the top plot to be "Diff. in percentage points".

**Comment by rickecon** [comment]
Yes. That looks really good. I can see some things we'll need to change.
1. Label the `x`-axis as "Unadjusted gross income". I think we can get rid of the "($thousands)" because we have the "K" in the category labels.
2. Change the `y`-axis labels of the bottom plot to be in thousands. Let's try the "K notation there also and see how it looks. Otherwise, we could put a little parenthetical "(000s)" in the axis label.
3. In the bottom plot, instead of the `y`-axis labels of "Tax units with tax increase" and "Tax units with tax cut", it might make more sense to put those two descriptions as a legend with "orange = Tax units with tax increase" and "gray = Tax units with tax cut".
  a. If we do this, we could leave the `y`-axis label blank if we use the "K" notation in the axis tick lables.
  b. Or we could put one `y`-axis label of (thousands of tax units).
4. The negative values in the `y`-axis of the bottom figure are confusing. Let's include a heavier black line at zero then label the negative values as positives, even though they are really negative in the data (I think).
5. The top plot is misleading. It makes the TJCA look like a huge cut for the wealthy. It may still be, but a more informative plot would be "Difference in tax liability as a percent of total income (effective tax rate)".
  a. This change would require the `y`-axis label of the top plot to be "Diff. in percentage points".

**Comment by rickecon** [comment]
Yes. That looks really good. I can see some things we'll need to change.
1. Label the `x`-axis as "Unadjusted gross income". I think we can get rid of the "($thousands)" because we have the "K" in the category labels.
2. Change the `y`-axis labels of the bottom plot to be in thousands. Let's try the "K notation there also and see how it looks. Otherwise, we could put a little parenthetical "(000s)" in the axis label.
3. In the bottom plot, instead of the `y`-axis labels of "Tax units with tax increase" and "Tax units with tax cut", it might make more sense to put those two descriptions as a legend with "orange = Tax units with tax increase" and "gray = Tax units with tax cut".
  a. If we do this, we could leave the `y`-axis label blank if we use the "K" notation in the axis tick lables.
  b. Or we could put one `y`-axis label of (thousands of tax units).
4. The negative values in the `y`-axis of the bottom figure are confusing. Let's include a heavier black line at zero then label the negative values as positives, even though they are really negative in the data (I think).
5. The top plot is misleading. It makes the TJCA look like a huge cut for the wealthy. It may still be, but a more informative plot would be "Difference in tax liability as a percent of total income (effective tax rate)".
  a. This change would require the `y`-axis label of the top plot to be "Diff. in percentage points".

**Comment by hayleefay** [comment]
@rickecon 
Current chart:
![image](https://user-images.githubusercontent.com/12599305/32570911-59f342f0-c48b-11e7-8355-4bf4b1d73a1c.png)

Notes:
**2**. I went with millions instead of K because it allowed me to use Bokeh functionality instead of writing a custom override of the labels and because I find '5m' more intuitive than '5000K'.
**4**. The parentheses around the formerly negative numbers are again a result of Bokeh functionality, I do think it is intuitive to differentiate between the two labels (positive and negative).
**5**. The variables currently coming out of Tax-Calculator restricted me to using after tax income rather than total income.  These are the variables from Tax-Calculator (I am using 'Change as % of After-Tax Income'):
- 'All Tax Units',
- 'Tax Units with Tax Cut',
- 'Percent with Tax Cut',
- 'Tax Units with Tax Increase',
- 'Percent with Tax Increase',
- 'Average Tax Change',
- 'Total Tax Difference',
- 'Share of Overall Change',
- 'Change as % of After-Tax Income',
- '% Change in After-Tax Income'

I can probably write something to use pretax income, it just depends on what our priorities are. I can also write something to try to change the issues noted in **2** and **4**, depending on priorities again.

**Comment by hayleefay** [comment]
@rickecon 
Current chart after adding the liability change bar chart:
![image](https://user-images.githubusercontent.com/12599305/32577081-63cfec3e-c49e-11e7-9780-a82164314375.png)

**Comment by hayleefay** [comment]
@rickecon And here is third vis when the data is binned by age.
![image](https://user-images.githubusercontent.com/12599305/32585367-bc010126-c4c2-11e7-97f2-71249ddd396d.png)

**Comment by rickecon** [comment]
@hayleefay . So so good. Comments:
1. How did you get 10 years (2018 to 2027) in the "Net change in government tax revenues" chart? (It should be "revenues" not "liabilities". My bad.) I thought Tax-Calculator was hard coded to stop at 2026.
2. I'm OK with the parentheses around the "Tax units with tax cut" numbers. Not ideal, but better than previous.
3. OK to use after tax income, if that is the easiest thing to get from Tax-Calculator. But one could reconstruct pre-tax income by adding tax liability to after tax income. But I am not sure what that number is relative to unadjusted gross income.
4. The plot in the top panel should have the title "Difference (in percentage points) in effective tax rates". That series is constructed in the following way. Define the following variables for each individual in the PUF data.
`tot_tax_liab_b`: total tax liability in the baseline
`tot_tax_liab_r`: total tax liability in the reform
`tot_inc`: total income in the baseline (this will be the same in baseline and reform), probably after-tax income.
`ETR_b` = `tot_tax_liab_b` / `tot_inc`: effective tax rate in the baseline
`ETR_r` = `tot_tax_liab_r` / `tot_inc`: effective tax rate in the reform
`ETR_diff` = `ETR_r` - `ETR_b`: Difference (in percentage points) in effective tax rates. Note that this is not a percentage change but rather a simple difference. When the two series being compared are already in percentage form, it is often best to just do a simple difference and call it "percentage point difference".

The series that I want to plot in the top panel of those figures is the `ETR_diff` average for each income bin (Figure 1) and each age bin (Figure 2), respectively.

**Comment by hayleefay** [comment]
@rickecon 
1. As of taxcalc 0.13.0, we can now extrapolate out to 2027 with the PUF. So I was able to just update taxcalc and then get 10 years of data.
2. I was able to write some js to take care of the parentheses so they are no longer there.

Working on 4 now.

**Comment by hayleefay** [comment]
@rickecon 4. Here is the code I used to create the simple difference.  
```
ETR_b = res1['combined']/res1['aftertax_income']
ETR_r = res2['combined']/res2['aftertax_income']
res2['perc_aftertax'] = ETR_r - ETR_b
```
There is also an `expanded_income` measure I can use rather than `aftertax_income` if you would like that instead.
This yielded the following percentage point differences for age: [-1.22, -1.29, -1.15, -1.24, -0.47] and the graph looks like this:
![image](https://user-images.githubusercontent.com/12599305/32613376-5c1e90fc-c530-11e7-8925-0709b4dff7f7.png)
And for income:
![image](https://user-images.githubusercontent.com/12599305/32613387-66e0564c-c530-11e7-9394-694004475cf6.png)

**Comment by rickecon** [comment]
Yes. Those look right to me. Let's make sure we have a label for the `y`-axis on both plots (e.g., "percentage point diff." Let's make the upper bound on both plots be +0.1, the lower bound on the age plot be -1.5, and the lower bound on the income plot be -2.0 (basically adding about 1/4 of empty space on the bottom). I can start writing with this stuff.

Another consideration that might really make those plots sing is to include shaded 25%-75% percentile bands, in addition to the line through the averages. This would give us some idea of the distribution of the tax change for each group without going to my dreaded box-whisker plot.
Below is an example.
![image](https://user-images.githubusercontent.com/4317690/32614327-0363ad8e-c53b-11e7-96da-6f8e426393af.png)

**Comment by hayleefay** [comment]
@rickecon Also, you might consider stacking the tax increase and tax cuts units (since they are both counting positive tax units):
![image](https://user-images.githubusercontent.com/12599305/32613964-ce1b1274-c531-11e7-91d2-de4a50465eee.png)
Or we can put them next to each other unstacked but both on the positive y-axis:
![image](https://user-images.githubusercontent.com/12599305/32614937-fa864a84-c533-11e7-97aa-67b204dd23e7.png)

**Comment by hayleefay** [comment]
@rickecon About the percentile bands, the function that is calculating the value for each bin is the `weighted_mean` function in taxcalc. It receives the data for each age or income bin and calculates the weighted mean for the bin, using s006 as the weights.
```
def weighted_mean(pdf, col_name):
    """
    Return weighted mean of Pandas DataFrame col_name items.
    """
    return ((pdf[col_name] * pdf['s006']).zsum() /
            (pdf['s006'].zsum() + EPSILON))
```

I am thinking of calculating the weighted percentiles by taking the bin data multipiled by the weights and then using numpy to find the percentiles:
```
data = pdf[col_name] * pdf['s006']
np.nanpercentile(data, 75)
np.nanpercentile(data, 25)
```
I'm not sure the results I am getting are making sense though:
Under 21: mean=-0.012247205250670776
75: 0.0
25: -19.7614436148

21-35: mean=-0.012905841003974631
75: 0.0
25: -19.0235036757

36-45: mean=-0.011538315630020914
75: 0.134490131756
25: -14.0743444837

46-65: mean=-0.012429316349318206
75: 0.0
25: -13.233959547

Over 65: mean=-0.004717745765278703
75: 0.0
25: -8.27770213594

**Comment by hayleefay** [comment]
@rickecon If you want to see how things change over time, here are the dynamic versions of age and income:
https://damp-mesa-27707.herokuapp.com/age_tcja
https://damp-mesa-27707.herokuapp.com/income_tcja

(These aren't on the homepage so the only way to access them is to use the link directly.)

**Comment by rickecon** [comment]
@hayleefay . Let's scrap the percentile bands idea for now. I think I like this version of the tax increase/cut plot.
<img width="606" alt="screen shot 2017-11-09 at 3 49 03 pm" src="https://user-images.githubusercontent.com/4317690/32628979-a8b2c250-c565-11e7-96d8-e0b713b1d808.png">

**Comment by rickecon** [comment]
I am writing right now.
